Index: flip2.c
===================================================================
--- flip2.c	(revision 1425)
+++ flip2.c	(working copy)
@@ -136,6 +136,7 @@
 static void flip2_display(PROGRAMMER* pgm, const char *prefix);
 static int flip2_program_enable(PROGRAMMER* pgm, AVRPART *part);
 static int flip2_chip_erase(PROGRAMMER* pgm, AVRPART *part);
+static int flip2_start_app(PROGRAMMER *pgm);
 static int flip2_read_byte(PROGRAMMER* pgm, AVRPART *part, AVRMEM *mem,
   unsigned long addr, unsigned char *value);
 static int flip2_write_byte(PROGRAMMER* pgm, AVRPART *part, AVRMEM *mem,
@@ -145,6 +146,7 @@
 static int flip2_paged_write(PROGRAMMER* pgm, AVRPART *part, AVRMEM *mem,
   unsigned int page_size, unsigned int addr, unsigned int n_bytes);
 static int flip2_read_sig_bytes(PROGRAMMER* pgm, AVRPART *part, AVRMEM *mem);
+static int flip2_parseexitspecs(PROGRAMMER* pgm, char *s);
 static void flip2_setup(PROGRAMMER * pgm);
 static void flip2_teardown(PROGRAMMER * pgm);
 
@@ -193,6 +195,7 @@
   pgm->read_byte        = flip2_read_byte;
   pgm->write_byte       = flip2_write_byte;
   pgm->read_sig_bytes   = flip2_read_sig_bytes;
+  pgm->parseexitspecs   = flip2_parseexitspecs;
   pgm->setup            = flip2_setup;
   pgm->teardown         = flip2_teardown;
 }
@@ -323,6 +326,10 @@
 void flip2_close(PROGRAMMER* pgm)
 {
   if (FLIP2(pgm)->dfu != NULL) {
+    if (pgm->exit_reset == EXIT_RESET_ENABLED) {
+        flip2_start_app(pgm);
+    }
+    
     dfu_close(FLIP2(pgm)->dfu);
     FLIP2(pgm)->dfu = NULL;
   }
@@ -388,6 +395,22 @@
   return cmd_result;
 }
 
+int flip2_start_app(PROGRAMMER *pgm) {
+    avrdude_message(MSG_INFO, "%s: Starting application\n", progname);
+    
+    struct flip2_cmd cmd = {
+        FLIP2_CMD_GROUP_EXEC, FLIP2_CMD_START_APP, { 0x00, 0, 0, 0 }
+    };
+    
+    // queue command
+    int cmd_result = dfu_dnload(FLIP2(pgm)->dfu, &cmd, sizeof(cmd));
+    
+    // repeat dnload to actually execute
+    dfu_dnload(FLIP2(pgm)->dfu, &cmd, sizeof(cmd));
+    
+    return cmd_result;
+}
+
 int flip2_read_byte(PROGRAMMER* pgm, AVRPART *part, AVRMEM *mem,
   unsigned long addr, unsigned char *value)
 {
@@ -502,6 +525,22 @@
   return (result == 0) ? n_bytes : -1;
 }
 
+/*
+ *  Parse the -E option string
+ */
+int flip2_parseexitspecs(PROGRAMMER* pgm, char *s)
+{
+    if (strcmp(s, "reset") == 0) {
+        pgm->exit_reset = EXIT_RESET_ENABLED;
+    } else if (strcmp(s, "noreset") == 0) {
+        pgm->exit_reset = EXIT_RESET_DISABLED;
+    } else {
+        return -1;
+    }
+    
+    return 0;
+}
+
 int flip2_read_sig_bytes(PROGRAMMER* pgm, AVRPART *part, AVRMEM *mem)
 {
   if (FLIP2(pgm)->dfu == NULL)
